File: python/nofpu_converter.py
""" ============================================================
// NoFPU Decimal Converter
// Developed by: Benyamin Gharri (Gharri.ir)
// Version: 1.0
// Python Port
// 
// This converter performs decimal conversion without using 
// Floating Point Unit (FPU)
// Ideal for embedded systems, microcontrollers, and systems 
// without FPU support
// ============================================================ """

"""
Convert to decimal format without using FPU
Suitable for systems without Floating Point Unit
"""

def convert_no_fpu(input_string):
    """
    Convert string to decimal format without using FPU
    - No float
    - No floating point operations
    - Only string and integer operations
    """
    # Remove spaces
    s = ''
    for ch in input_string:
        if ch != ' ':
            s += ch
    
    # Find '-' with simple loop
    dash_pos = -1
    for i in range(len(s)):
        if s[i] == '-':
            dash_pos = i
            break
    
    if dash_pos == -1:
        return None, "ERROR: '-' not found"
    
    # Extract parts
    dec_str = s[:dash_pos]
    num_str = s[dash_pos + 1:]
    
    # Validation
    if not dec_str:
        return None, "ERROR: decimal count is empty"
    
    # Manual conversion of decimal count to integer (without int())
    dec_count = 0
    for ch in dec_str:
        if '0' <= ch <= '9':
            dec_count = dec_count * 10 + (ord(ch) - 48)
        else:
            return None, "ERROR: invalid decimal count"
    
    if not num_str:
        return None, "ERROR: number part is empty"
    
    # If decimal count is zero
    if dec_count == 0:
        return num_str, None
    
    # Padding if needed (with minimal memory)
    if dec_count > len(num_str):
        zeros_needed = dec_count - len(num_str)
        num_str = ('0' * zeros_needed) + num_str
    
    # Separate integer and decimal parts
    if dec_count >= len(num_str):
        result = "0." + num_str
    else:
        split_pos = len(num_str) - dec_count
        result = num_str[:split_pos] + "." + num_str[split_pos:]
    
    return result, None


# Ultra-lightweight version for microcontrollers
def convert_micro(input_string):
    """
    Microcontroller-specific version with minimal memory usage
    """
    result = []
    dec_count = 0
    num_part = ""
    is_dec = True
    found_dash = False
    
    # Single pass through string (no extra memory allocation)
    for ch in input_string:
        if ch == ' ':
            continue
        elif ch == '-':
            found_dash = True
            is_dec = False
        elif is_dec:
            if '0' <= ch <= '9':
                dec_count = dec_count * 10 + (ord(ch) - 48)
            else:
                return "ERROR"
        else:
            if '0' <= ch <= '9':
                num_part += ch
            else:
                return "ERROR"
    
    if not found_dash or not num_part:
        return "ERROR"
    
    if dec_count == 0:
        return num_part
    
    if dec_count > len(num_part):
        num_part = ('0' * (dec_count - len(num_part))) + num_part
    
    if dec_count >= len(num_part):
        return "0." + num_part
    else:
        split_pos = len(num_part) - dec_count
        return num_part[:split_pos] + "." + num_part[split_pos:]


# Management class for systems without FPU
class NoFPUConverter:
    """
    Management class for conversion on systems without FPU
    """
    def __init__(self):
        self.cache = {}
        self.total_conversions = 0
    
    def convert(self, input_str):
        """Convert with result caching"""
        if input_str in self.cache:
            return self.cache[input_str]
        
        result = convert_micro(input_str)
        
        if result != "ERROR" and len(self.cache) < 100:
            self.cache[input_str] = result
        
        self.total_conversions += 1
        return result
    
    def clear_cache(self):
        """Clear cache to free memory"""
        self.cache.clear()
    
    def get_stats(self):
        """Get conversion statistics"""
        return {
            'total': self.total_conversions,
            'cache_size': len(self.cache)
        }


# Main program
if __name__ == "__main__":
    print("=" * 60)
    print("NoFPU Decimal Converter - Suitable for embedded systems")
    print("=" * 60)
    print("\nFeatures:")
    print("✓ No float usage")
    print("✓ No floating point operations")
    print("✓ Minimal CPU usage")
    print("✓ Minimal memory usage")
    print("✓ Suitable for systems without FPU")
    print("\nInput format: decimal_count-number")
    print("Example: 2-101013 -> 1010.13")
    print("Enter 'exit' to quit\n")
    
    converter = NoFPUConverter()
    
    while True:
        user_input = input("Input: ").strip()
        
        if user_input.lower() == 'exit':
            print("\nStatistics:")
            stats = converter.get_stats()
            print(f"Total conversions: {stats['total']}")
            print(f"Cache size: {stats['cache_size']}")
            print("Exiting program")
            break
        
        if not user_input:
            print("Please enter a value\n")
            continue
        
        result = converter.convert(user_input)
        
        if result == "ERROR":
            print("❌ Error: Invalid input format")
            print("Correct format: decimal_count-number (Example: 2-101013)\n")
        else:
            print(f"✅ Result: {result}")
            print(f"   (Without using FPU)\n")

File: Go/nofpu_converter.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// ============================================================
// NoFPU Decimal Converter
// Developed by: Benyamin Gharri (Gharri.ir)
// Version: 1.0
// Go Port
// 
// This converter performs decimal conversion without using 
// Floating Point Unit (FPU)
// Ideal for embedded systems, microcontrollers, and systems 
// without FPU support
// ============================================================

// convertNoFPU converts string to decimal format without using FPU
// - No float
// - No floating point operations
// - Only string and integer operations
func convertNoFPU(inputString string) (string, string) {
	// Remove spaces
	s := ""
	for _, ch := range inputString {
		if ch != ' ' {
			s += string(ch)
		}
	}

	// Find '-' with simple loop
	dashPos := -1
	for i, ch := range s {
		if ch == '-' {
			dashPos = i
			break
		}
	}

	if dashPos == -1 {
		return "", "ERROR: '-' not found"
	}

	// Extract parts
	decStr := s[:dashPos]
	numStr := s[dashPos+1:]

	// Validation
	if decStr == "" {
		return "", "ERROR: decimal count is empty"
	}

	// Manual conversion of decimal count to integer (without strconv.Atoi)
	decCount := 0
	for _, ch := range decStr {
		if ch >= '0' && ch <= '9' {
			decCount = decCount*10 + int(ch-'0')
		} else {
			return "", "ERROR: invalid decimal count"
		}
	}

	if numStr == "" {
		return "", "ERROR: number part is empty"
	}

	// If decimal count is zero
	if decCount == 0 {
		return numStr, ""
	}

	// Padding if needed
	if decCount > len(numStr) {
		zerosNeeded := decCount - len(numStr)
		numStr = strings.Repeat("0", zerosNeeded) + numStr
	}

	// Separate integer and decimal parts
	var result string
	if decCount >= len(numStr) {
		result = "0." + numStr
	} else {
		splitPos := len(numStr) - decCount
		result = numStr[:splitPos] + "." + numStr[splitPos:]
	}

	return result, ""
}

// convertMicro is the ultra-lightweight version for microcontrollers
func convertMicro(inputString string) string {
	var result []rune
	decCount := 0
	numPart := ""
	isDec := true
	foundDash := false

	// Single pass through string (no extra memory allocation)
	for _, ch := range inputString {
		if ch == ' ' {
			continue
		} else if ch == '-' {
			foundDash = true
			isDec = false
		} else if isDec {
			if ch >= '0' && ch <= '9' {
				decCount = decCount*10 + int(ch-'0')
			} else {
				return "ERROR"
			}
		} else {
			if ch >= '0' && ch <= '9' {
				numPart += string(ch)
			} else {
				return "ERROR"
			}
		}
	}

	if !foundDash || numPart == "" {
		return "ERROR"
	}

	if decCount == 0 {
		return numPart
	}

	if decCount > len(numPart) {
		numPart = strings.Repeat("0", decCount-len(numPart)) + numPart
	}

	var resultStr string
	if decCount >= len(numPart) {
		resultStr = "0." + numPart
	} else {
		splitPos := len(numPart) - decCount
		resultStr = numPart[:splitPos] + "." + numPart[splitPos:]
	}

	_ = result // just to avoid unused variable warning
	return resultStr
}

// NoFPUConverter is the management class for systems without FPU
type NoFPUConverter struct {
	cache            map[string]string
	totalConversions int
}

// NewNoFPUConverter creates a new converter instance
func NewNoFPUConverter() *NoFPUConverter {
	return &NoFPUConverter{
		cache:            make(map[string]string),
		totalConversions: 0,
	}
}

// Convert performs conversion with result caching
func (c *NoFPUConverter) Convert(inputStr string) string {
	// Check cache
	if val, exists := c.cache[inputStr]; exists {
		return val
	}

	// Convert using micro version
	result := convertMicro(inputStr)

	// Cache if not error and cache size < 100
	if result != "ERROR" && len(c.cache) < 100 {
		c.cache[inputStr] = result
	}

	c.totalConversions++
	return result
}

// ClearCache clears cache to free memory
func (c *NoFPUConverter) ClearCache() {
	c.cache = make(map[string]string)
}

// GetStats gets conversion statistics
func (c *NoFPUConverter) GetStats() (total int, cacheSize int) {
	return c.totalConversions, len(c.cache)
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("NoFPU Decimal Converter - Suitable for embedded systems")
	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("\nFeatures:")
	fmt.Println("✓ No float usage")
	fmt.Println("✓ No floating point operations")
	fmt.Println("✓ Minimal CPU usage")
	fmt.Println("✓ Minimal memory usage")
	fmt.Println("✓ Suitable for systems without FPU")
	fmt.Println("\nInput format: decimal_count-number")
	fmt.Println("Example: 2-101013 -> 1010.13")
	fmt.Println("Enter 'exit' to quit\n")

	converter := NewNoFPUConverter()

	for {
		fmt.Print("Input: ")
		scanner.Scan()
		userInput := strings.TrimSpace(scanner.Text())

		if strings.ToLower(userInput) == "exit" {
			fmt.Println("\nStatistics:")
			total, cacheSize := converter.GetStats()
			fmt.Printf("Total conversions: %d\n", total)
			fmt.Printf("Cache size: %d\n", cacheSize)
			fmt.Println("Exiting program")
			break
		}

		if userInput == "" {
			fmt.Println("Please enter a value\n")
			continue
		}

		result := converter.Convert(userInput)

		if result == "ERROR" {
			fmt.Println("❌ Error: Invalid input format")
			fmt.Println("Correct format: decimal_count-number (Example: 2-101013)\n")
		} else {
			fmt.Printf("✅ Result: %s\n", result)
			fmt.Println("   (Without using FPU)\n")
		}
	}
}